### ISSUE TITLE
Possible memory leak in model.train() with stateful RNN and WebGL backend only

### ISSUE THREAD

**Issue by brannondorsey** [tensorflow/tfjs#664]
#### TensorFlow.js version

tfjs v0.12.6

**update:** have also now tested with tfjs v0.12.7 as well. The problem persists.

#### Browser version

Tested on Chrome and Firefox latest, as well as Electron v2.0.8

#### Describe the problem or feature request

I'm importing [this stateful Keras RNN model](https://github.com/brangerbriz/twitter-transfer-learning/tree/578999703d19b83348c0aa5d2a6a4592852da73f/checkpoints/base-model-10M/tfjs) into tfjs for further training/fine tuning. The model training behaves as expected when using the `tensorflow` backend with `tfjs-node` (CPU) and `tfjs-node-gpu` (CUDA) as well as the `cpu` browser backend. However, when the `webgl` backend is used, `model.fit(x, y, options)` causes a catastrophic memory leak ([here](https://github.com/brangerbriz/twitter-transfer-learning/blob/578999703d19b83348c0aa5d2a6a4592852da73f/src/main.js#L53-L61) is the specific line that appears to be causing the leak).

![screenshot from 2018-08-31 15-04-06](https://user-images.githubusercontent.com/3193533/44934170-d8646780-ad39-11e8-9d05-66d997cab052.png)
![screenshot from 2018-08-31 15-05-38](https://user-images.githubusercontent.com/3193533/44934177-e0240c00-ad39-11e8-90b2-390b2d0da43e.png)

Again, this appears to only be a problem with the WebGL backend and is not experienced at all with cpu training or `tfjs-node-gpu`. Occasionally I get WebGL warning logged to the console like so, but this does not always occur:

![screenshot from 2018-08-31 13-56-20](https://user-images.githubusercontent.com/3193533/44934419-bae3cd80-ad3a-11e8-8e26-097fad41f46b.png)

I've also seen what appears to be a loss of the WebGL context, with a juicy [fragment shader error here](https://pastebin.com/KVBXjRpt). I've also noticed that in addition to an exhaustion of system memory as seen above, GPU memory also appears to be quickly exhausted. Monitoring GPU memory usage in realtime with `watch -n 0.1 nvidia-smi` shows cycles where memory rapidly increases, plateaus at maximum usage for a while, and then rapidly decreases until it's almost completely released before starting the cycle again. 

Here is a description of [the model](https://github.com/brangerbriz/twitter-transfer-learning/tree/578999703d19b83348c0aa5d2a6a4592852da73f/checkpoints/base-model-10M/tfjs) being trained:
```javascript
Sequential:
Embedding // 32
Dropout // 0.2
LSTM // 256 units, stateful = true
Tanh
Dropout // 0.2
LSTM // 256 units, stateful = true
Tanh
Dropout // 0.2
TimeDistributedDense // 98 output classes trained w/ categorical crossentropy loss
```
This is only speculation, but I wonder if the leak may have something to do with the new support for stateful RNNs with https://github.com/tensorflow/tfjs-layers/pull/292

#### Code to reproduce the bug / link to feature request

I've made a [repository available](https://github.com/brangerbriz/twitter-transfer-learning/tree/tfjs-issue-664) so that maintainers can experience this possible memory leak themselves. Please be sure to use the `tfjs-issue-664` branch.

```bash
# clone the repo using the branch for this issue
git clone https://github.com/brangerbriz/twitter-transfer-learning
git checkout -b tfjs-issue-664 && git pull origin tfjs-issue-664
cd twitter-transfer-learning

# install deps
npm install

# open up a system monitor to view your CPU's memory usage before launching...

# launch an electron app that loads the keras model and begins to fine tune it with model training
npm start
```

You will notice that the leak begins after `model.fit()`'s [`onBatchBegin` callback](https://github.com/brangerbriz/twitter-transfer-learning/blob/578999703d19b83348c0aa5d2a6a4592852da73f/src/main.js#L59), but before `onBatchEnd` is called. The batch sizes I am using during training are not outrageous, I've tested at 128, 64, and 32, but the problem persists. I should also mention that I am using a GTX1080 GPU.

I'm happy to provide more details or further explain bits the code if necessary. Thanks!

**Comment by tafsiri** [comment]
Thanks for the report @brannondorsey. @caisq could you take a look at this and give your thoughts on whether the issue may be in stateful RNN?

**Comment by brannondorsey** [comment]
Any thoughts on this?

**Comment by brannondorsey** [comment]
I've made a new branch called [`tfjs-issue-664`](https://github.com/brangerbriz/twitter-transfer-learning/tree/tfjs-issue-664) in the `twitter-transfer-learning` repository mentioned in the original issue. I've also updated all links in this issue to point to commits in the git tree so that they will be preserved as new commits are added to the repository's `master` branch.

**Comment by caisq** [comment]
@brannondorsey FYI, it is on my TODO list to look at this. But because of my schedule, the effort will be delay for 1-2 weeks. Thanks for your understanding.

**Comment by brannondorsey** [comment]
@caisq, gotcha, thanks for the update. Please let me know if you have any questions or would like any further information from me.

**Comment by rthadur** [comment]
Closing this request as we are more focussed on loading pre trained models not the training the models anymore. Please @me if you still need this request, I will reopen.Thank you
